lib/color_spinor_wuppertal.cu

```diff
@@ -0,0 +1,296 @@
+                       Float A, Float B)
+      : out(out), in(in), U(U), A(A), B(B), parity(parity), nParity(in.SiteSubset()), nFace(1),
+        dim{ (3-nParity) * in.X(0), in.X(1), in.X(2), in.X(3), 1 },
+      commDim{comm_dim_partitioned(0), comm_dim_partitioned(1), comm_dim_partitioned(2), comm_dim_partitioned(3)},
+      volumeCB(in.VolumeCB())
+    {
+      if (in.FieldOrder() != QUDA_FLOAT2_FIELD_ORDER || !U.isNative())
+        errorQuda("Unsupported field order colorspinor=%d gauge=%d combination\n", in.FieldOrder(), U.FieldOrder());
+    }
+  };
+
+  /**
+     Computes out = sum_mu U_mu(x)in(x+d) + U^\dagger_mu(x-d)in(x-d)
+     @param[out] out The out result field
+     @param[in] U The gauge field
+     @param[in] in The input field
+     @param[in] x_cb The checkerboarded site index
+     @param[in] parity The site parity
+  */
+  template <typename Float, int Nc, typename Vector, typename Arg>
+  __device__ __host__ inline void computeNeighborSum(Vector &out, Arg &arg, int x_cb, int parity) {
+
+    typedef Matrix<complex<Float>,Nc> Link;
+    const int their_spinor_parity = (arg.nParity == 2) ? 1-parity : 0;
+
+    int coord[5];
+    getCoords(coord, x_cb, arg.dim, parity);
+    coord[4] = 0;
+
+#pragma unroll
+    for (int dir=0; dir<3; dir++) { // loop over spatial directions
+
+      //Forward gather - compute fwd offset for vector fetch
+      const int fwd_idx = linkIndexP1(coord, arg.dim, dir);
+
+      if ( arg.commDim[dir] && (coord[dir] + arg.nFace >= arg.dim[dir]) ) {
+        const int ghost_idx = ghostFaceIndex<1>(coord, arg.dim, dir, arg.nFace);
+
+        const Link U = arg.U(dir, x_cb, parity);
+	const Vector in = arg.in.Ghost(dir, 1, ghost_idx, their_spinor_parity);
+
+        out += U * in;
+      } else {
+        const Link U = arg.U(dir, x_cb, parity);
+	const Vector in = arg.in(fwd_idx, their_spinor_parity);
+
+        out += U * in;
+      }
+
+      //Backward gather - compute back offset for spinor and gauge fetch
+      const int back_idx = linkIndexM1(coord, arg.dim, dir);
+      const int gauge_idx = back_idx;
+
+      if ( arg.commDim[dir] && (coord[dir] - arg.nFace < 0) ) {
+        const int ghost_idx = ghostFaceIndex<0>(coord, arg.dim, dir, arg.nFace);
+
+        const Link U = arg.U.Ghost(dir, ghost_idx, 1-parity);
+	const Vector in = arg.in.Ghost(dir, 0, ghost_idx, their_spinor_parity);
+
+        out += conj(U) * in;
+      } else {
+        const Link U = arg.U(dir, gauge_idx, 1-parity);
+	const Vector in = arg.in(back_idx, their_spinor_parity);
+
+        out += conj(U) * in;
+      }
+    }
+  }
+
+  //out(x) = A in(x) + B computeNeighborSum(out, x)
+  template <typename Float, int Ns, int Nc, typename Arg>
+  __device__ __host__ inline void computeWupperalStep(Arg &arg, int x_cb, int parity)
+  {
+    typedef ColorSpinor<Float,Nc,Ns> Vector;
+    Vector out;
+
+    computeNeighborSum<Float,Nc>(out, arg, x_cb, parity);
```

Comment:
Continuing from the other comment: Some sort of barrier ( __threadfence() ) is placed after the result is calculated in full, then broadcast. Rather than calling this entire routine N times for an N step operation, the routine would be called once, the links gathered once, then the vectors would be computed N times within the function.